contrib/masterfiles-stage/common.sh

```diff
@@ -12,146 +12,217 @@ if [ ! "$debug_mode" == "true" ]; then
-      fi
-      rm -rf "${STAGING_DIR}"/* "${STAGING_DIR}"/.??*
-      if [ "$debug_mode" == "true" ]; then
-        echo "Cloning git repository '${GIT_URL}' '${GIT_BRANCH}' branch into '${STAGING_DIR}'"
-      fi
-      git clone --no-hardlinks "${GIT_URL}" "${STAGING_DIR}"
-
-      if [ "$debug_mode" == "true" ]; then
-        echo "Checking out '${GIT_BRANCH}' branch."
-      fi
-      cd "${STAGING_DIR}" && git checkout "${GIT_BRANCH}"
-  fi
+  STAGING_DIR="${ROOT}/_tmp$(echo "$PARAMS" | tr [./] _)"
+}
 
-  if /var/cfengine/bin/cf-promises -T "${STAGING_DIR}"; then
-    if ! /var/cfengine/bin/cf-promises -cf "${STAGING_DIR}/update.cf"; then
-        error_exit "Update policy staged in ${STAGING_DIR} could not be validated, aborting."
-    fi
+check_git_installed() {
+  git --version >/dev/null 2>&1 || error_exit "git not found on path: '${PATH}'"
+}
 
-      # you could abort here if DIFFLINES is over 100, for instance (too many changes)
-      #DIFFLINES=$(/usr/bin/diff -r  -x .git -x cf_promises_validated -x cf_promises_release_id "${STAGING_DIR}" "${MASTERDIR}" |/usr/bin/wc -l)
-
-      # roll out the release if the release IDs are different
-      # ALWALSY SYNC THEM but what is the negative side effect? POTENTIALY CLIENTS ALWWAYS UPDATE?
-      # BUT CPV only triggered on change, so maybe thats what we want.
-      #if /usr/bin/diff -q "${STAGING_DIR}/cf_promises_release_id" "${MASTERDIR}/cf_promises_release_id" ; then
-      #    echo "No release needs to be made, the release IDs are the same."
-      #    touch "${STAGING_DIR}"
-      #else
-          /bin/mkdir -p "${MASTERDIR}" || error_exit "Failed: Creating '${MASTERDIR}'"
-          cd "${STAGING_DIR}" && (
-          chown -R root:root "${STAGING_DIR}" && \
-          rsync -rltDE -c --delete-after --chmod=u+rwX,go-rwx "${STAGING_DIR}/" "${MASTERDIR}/" && echo "Successfully deployed branch '${GIT_BRANCH}' from '${GIT_URL}' to '${MASTERDIR}' on $(date)"
-      )
-      #fi
-  else
-    error_exit "The staged policies in ${STAGING_DIR} could not be validated, aborting."
+git_setup_local_mirrored_repo() {
+  # This function sets the variable local_mirrored_repo to a directory path
+  # based on the value of GIT_URL and ROOT, and if that directory doesn't exist,
+  # creates it as a mirrored clone of the repo at GIT_URL.  If it does exist,
+  # update it with git fetch.
+  #
+  # This code could be improved if there is an inexpensive way to check that
+  # a local bare repository is in fact a *mirrored* repository of a specified
+  # GIT_URL, but for now if the local_mirrored_repo is in fact a bare git
+  # repo (guaranteed by the success of the "git fetch" command) then we just
+  # assume it is a *mirrored* repository.
+  #
+  # Since the pathname is directly based on GIT_URL, there is no chance
+  # of *accidental* name collision.
+
+  local_mirrored_repo="${ROOT}/$(printf '%s' "${GIT_URL}" | sed 's/[^A-Za-z0-9._-]/_/g')"
```

Comment:
@nickanderson if you work out why putting stuff in ROOT breaks Design Center I'd be happy to change this.  Otherwise I think I should just add a new parameter "staging_dirs_container" or something like that (to params.sh), and use it the way I've been using ROOT.

I understand ROOT needs to be *in* `params.sh` to integrate with Design Center; I don't think that means I have to use it at all for the staging script.  I originally used it this way only because I thought that was the *purpose* of the ROOT parameter.